### ISSUE TITLE
About the cost of research...

### ISSUE THREAD

**Issue by KilloZapit** [pahimar/Equivalent-Exchange-3#1002]
I personally think one of the problems with this mod as it exists now is that researching things doesn't really require that big of an investment most of the time. It only requires one of the item you want to research and a very tiny amount of time.

I kinda think it should cost experience levels instead, maybe a number of levels based on the binary logarithm (reverse powers of two, since a lot off the EMC values of objects seem to be based on powers of two) and maybe a multiplier based on complexity (basically it costs more the more crafting steps removed from the raw materials it is, which should be easy to calculate during the EMC profiling when the results of crafting recipes are given EMC values).

That, I think, would make it so becoming proficient in transmutation is an investment. Anyone could  dump items into a table and convert things to EMC, but only someone who spent the experience could use EMC from items to make things. I don't know if it would be more balanced exactly, but I think it might be more interesting.

But that's just my silly idea.

**Comment by mitchfizz05** [comment]
I can certainly agree that researching items is far to easy, and an extra cost should be intergrated, be it experience points or something else.

**Comment by x3n0ph0b3** [comment]
My original suggestion was a floored fourth root, but the numbers seemed a bit high toward the advanced end of the spectrum, with the level system in minecraft only allowing for values at or below 6.25 million. (50 * 50 * 50 * 50).

A better system would probably be a Math.log(emcValue), allowing for emcValues of 1 to cost nothing [trivial] while anything above 1 would be a fraction until exceeding 1. For added style points, fractions of the exp bar could be valid (instead of flooring it), because experience in minecraft is stored as a float. It would give better representation to those low-values while heavily curbing the "extreme" value spectrum to something reasonable.

Roughly dark matter would cost around 5.77 levels to learn.
A diamond would take around 3.91.

Using a Math.log, it would take a ludicrously inflated number to exceed Minecraft's experience cap. I'm pretty sure it's the best tool for the job in this scenario, if experience costs were going to be a thing.

**Comment by KilloZapit** [comment]
Yeah, my original idea was a diamond would cost 13 levels since 2^13 = 8192.

**Comment by x3n0ph0b3** [comment]
Similar concept, powers of 2 seems high, but would cover bases. I'm not sure I can agree with the scale. 13 levels is a "lot" for a diamond. I think for a harder/more costly experience it's a good option. Personally I'd prefer the log 10, it's a bit more rounded down. Base 2 would definitely work though, imo.

**Comment by UndeadZeratul** [comment]
maybe take a page from Spice of Life's book and let the user define a custom formula?

**Comment by x3n0ph0b3** [comment]
+1, even though a base 2 isn't my personal pref, i can see a lot of maps/users wanting to define their own base/curve. Java has to deal with a "shortcoming" in that it only predefines 2 bases [as of java 7, unsure if 8 adds anything new]

Math.log is the equivalent of "e" (natural log) and math.log10 is for the other customary "default", base 10.

To arrive at a base other than 10 or e, one can use the identity:
                log[a]x
 log[b]x = ---------
                log[a]b

Thus it would be trivial to allow a configuration to handle "b", where a is simply one static log method (either java's default [e] or log10) and a is the EMC value of an item in question. The result would always be the configured base.

**Comment by KilloZapit** [comment]
I kinda think diamonds are on the high end of things you would want to learn myself, but yeah customization is a good idea.

**Comment by x3n0ph0b3** [comment]
My main concern would be mod content adding things at such a high end of the vanilla spectrum that they produce exorbitant sigmas of experience costs when looked at as a whole.

For example, thing takes 8 diamond blocks. [Dark matter, for reference]
Thing2 takes 9 of thing1. [Just an example]
Thing3 takes 9 of thing2. [Expounding]

At a log2, Thing3 costs 25 experience levels. At a log of 10, it costs around 7.
While one balance may be suitable for a given player, I find 25 levels to be steep for the knowledge to craft a single item. I think 7 is a bit more reasonable. 

It can't be overstated, it really boils down to player preference.

**Comment by WolfAmaril** [comment]
We could always throw in a config option to let players and server owners set the log base for different items.

**Comment by KilloZapit** [comment]
Funny, I never have a problem getting a hold of tons of exp, mostly from mining every bit of coal I see, which I have way more incentive to do with this mod. 25 levels just doesn't seem like that much to me. :P

**Comment by KilloZapit** [comment]
Is 25 levels really that much? I don't even farm and I tend to get that pretty easily after not too long.

**Comment by zauberparacelsus** [comment]
Bear in mind that experience levels are not equal in the amount of experience points.  When paying 5 levels, for example, you wind up paying 105 experience points when you are at level 10, but when are at level 25, you wind up paying 360 experience points.

This behavior is something that can (for example) be exploited if you have OpenBlocks installed, where if you need to pay experience levels for something, you use a drain to dump all of your experience levels, and then punch the storage tank until you have the required amount of levels you need to pay.  Doing this, you wind up paying much less experience than you would if you had just paid while you had all of your levels.

More info: http://minecraft.gamepedia.com/Experience#Leveling_up

**Comment by WolfAmaril** [comment]
So by that logic it would probably make more sense to have research cost experience points, as opposed to levels.

**Comment by KilloZapit** [comment]
I do use OpenBlocks myself... I always kinda figured experience costs were "cost enough experience to get to level x" though. The system seems kinda broken otherwise.

**Comment by x3n0ph0b3** [comment]
Agree that points make more sense than levels. In general though, I feel like "MC Experience" is generic and underwhelming. There is a suggestion on the research costs in the idea stack that I like much, much better than this one at this point.

**Comment by KilloZapit** [comment]
That did give me the idea that maybe you should have an alchemy experience that unlocks more options for research as you research simple stuff, but that doesn't exactly encourage specialization. That's why I wanted to use MC experience, because you would have to choose more what you wanted to spend it on.

**Comment by x3n0ph0b3** [comment]
Actually, I'm starting to bend on the idea of MC experience. It already has its own inherent self-policing "you can't get too strong at once" constraint and I can sort of get behind it. I didn't like it but I'm warming up to it, so let's not remove MC experience from consideration just yet - I was being too hasty.

I am now including our other discussion and attempting to incorporate it here and there. I feel like perhaps we need to move this to a unified format where we can see everything in one location, if such a thing exists.

**Comment by x3n0ph0b3** [comment]
I'm going to make a case for MC experience here that I don't think the way it is set up is without merit.

A lot of people probably felt the same way I did about MC experience at first.
"Why is each level worth slightly more than the last, that's dumb. It encourages you to stay at low levels, where experience goes farther, and you're spending less of it to accomplish the same goals"

To this I say "exactly". It is self policing, in that you can't get too strong, too fast. It's the point of the mechanic to prevent you from stockpiling a massive amount of raw exp and dumping it into things, but rather that it becomes progressively harder to accomplish effects which require higher levels of experience in a single sitting.

Guys, this is a callback to Dark Souls. For those of you who don't play Dark Souls, it's a grinding punishment mechanic. It is trying to encourage you to spend those points and keep moving, in a sort of roundabout way. Feel free to disagree here, but that's my take on the system.

**Comment by KilloZapit** [comment]
I think MC experience is a good system it's just not used for very much. Once you get the best armor you can with the best enchantments, that's it. I mean I guess you could make a lot of back up armor and stuff, but it still seemed to simple.

And I thought the "each level cost more then the last" was fairly obvious way of doing balance. But again, I always assumed when enchanting cost levels it actually just took the amount of experience to reach that level, and I am still not sure if that is true or not. Just taking x levels off your total makes it so you basically need to spend levels as soon as possible. I think I remember testing it at one point... Can't remember for sure the result.

**Comment by x3n0ph0b3** [comment]
It is true, it does not take the same amount of experience to reach level 1 as it does to reach level 50. I learned this while creating the Hero's Medallion in Reliquary, et al. Each level successively costs more than the last.
